underfs/b2/src/main/java/alluxio/underfs/b2/B2UnderFileSystem.java

```diff
@@ -0,0 +1,374 @@
+      for (B2FileVersion version : mB2StorageClient.fileVersions(request)) {
+        if (version.getFileName().equals(key)) {
+          mB2StorageClient.deleteFileVersion(version);
+        } else {
+          break;
+        }
+      }
+    } catch (B2Exception e) {
+      LOG.error("Failed to delete {}", key, e);
+      return false;
+    }
+    return true;
+  }
+
+  @Override
+  protected String getFolderSuffix() {
+    LOG.debug("getFolderSuffix {}", FOLDER_SUFFIX);
+    return FOLDER_SUFFIX;
+  }
+
+  @Nullable
+  @Override
+  protected ObjectListingChunk getObjectListingChunk(String key, boolean recursive)
+      throws IOException {
+    LOG.debug("getObjectListingChunk({}, {})", key, recursive);
+    String delimiter = recursive ? null : PATH_SEPARATOR;
+    key = PathUtils.normalizePath(key, PATH_SEPARATOR);
+    // In case key is root (empty string) do not normalize prefix.
+    key = key.equals(PATH_SEPARATOR) ? "" : key;
+
+    LOG.debug("key: {}", key);
+    LOG.debug("delimiter: {}", delimiter);
+
+    final B2ListFileNamesRequest.Builder builder =
+        B2ListFileNamesRequest.builder(mBucketId).setMaxFileCount(getListingChunkLength(mUfsConf))
+            .setPrefix(key);
+
+    if (delimiter != null) {
+      builder.setDelimiter(delimiter);
+    }
+
+    final B2ListFileNamesRequest request = builder.build();
+
+    try {
+      return new B2ObjectListingChunk(mB2StorageClient.fileNames(request));
+    } catch (B2Exception e) {
+      Throwables.propagateIfPossible(e, IOException.class);
+      return null;
+    }
+  }
+
+  @Nullable
+  @Override
+  protected ObjectStatus getObjectStatus(String key) throws IOException {
+    LOG.debug("bucketName: {}, key: {}", mBucketName, key);
+    try {
+      B2FileVersion fileInfo = mB2StorageClient.getFileInfoByName(mBucketName, key);
+      return new ObjectStatus(fileInfo.getFileName(), fileInfo.getContentSha1(),
+          fileInfo.getContentLength(), fileInfo.getUploadTimestamp());
+    } catch (B2Exception e) {
+      if (e.getCode().equals("not_found")) {
+        LOG.debug("key not found {}", key);
+        return null;
+      }
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  protected ObjectPermissions getPermissions() {
+    return new ObjectPermissions(mAccountId, mAccountId, mBucketMode);
+  }
+
+  @Override
+  protected String getRootKey() {
+    LOG.debug("getRootKey {}{}", B2_SCHEME, mBucketName);
+    return B2_SCHEME + mBucketName;
+  }
+
+  @Override
+  public String getUnderFSType() {
+    return "b2";
+  }
+
+  @Override
+  protected InputStream openObject(String key, OpenOptions options, RetryPolicy retryPolicy)
+      throws IOException {
+    LOG.debug("key [{}]", key);
+    LOG.debug("options [{}]", options);
+    try {
+      return new B2InputStream(mBucketName, key, mB2StorageClient, options.getOffset(),
+          options.getLength());
+    } catch (ServiceException e) {
+      throw new IOException(e.getMessage());
```

Comment:
Instead of repeating the initial exception message, can we add context to the message like `new IOException(String.format("Failed to open object: bucket: %s key: %s", mBucketName, key), e)`